### DISCUSSION TITLE
How other missions handle finite state machines in F'?

### DISCUSSION THREAD

**Discussion by sobkulir** [nasa/fprime#2090] [Q&A] [answered]
We're a group of students from ETH Zurich building a CubeSat and we're planning on using F' for our mission[0]. One problem which standard components seem not to address is how state of the CubeSat is handled. For example, for simplicity, there might be three states: {`Detumbling`, `Regeneration`, `Science`}.

Now, let's say a command to do some Science is being dispatched, for example, "Tell Payload to do Experiment A". This command shall go to a component that handles Payload logic, say `PayloadMgr`. The component that dispatches commands, `Svc::CmdDispatcher`, does not do any state checking and dispatches the command directly to `PayloadMgr`. This means that the `PayloadMgr` component either needs to:
1) Query the global state from a component that holds it, say `StateMgr`. Now either a) `StateMgr` needs to know about `PayloadMgr`'s commands and authorize it, or b) The `PayloadMgr` needs to know about states and decide which ones are okay to run in. 
or
2) Have a state regarding `PayloadMgr` internally stored in `PayloadMgr` component and decide whether to execute the command according to this state. It can be: {`NoOperation`, `ExperimentAStarted`, ... }. This way, the state is more distributed.

This sounds a lot like hierarchical state machines[1]. In the case 1) the `StateMgr` has all the state, while in 2) the substates are in different components.

Personally, I find approach 1b to be the most clean one. I would be very curious to know how other F Prime missions handled the state as it would simplify our design considerably. The paper on Lunar Flashlight[2] mentions `FP State Mgr` component, but it has not been open-sourced[3].

[0] https://aris-space.ch/sage-cubesat/
[1] https://en.wikipedia.org/wiki/UML_state_machine#Hierarchically_nested_states
[2] Developing Lunar Flashlight and Near-Earth Asteroid Scout Flight Software Concurrently
using Open-Source F Prime Flight Software Framework, https://digitalcommons.usu.edu/cgi/viewcontent.cgi?article=5263&context=smallsat
[3] https://github.com/fprime-community/fprime-sphinx/

**Comment by timcanham** [comment, marked as answer]
@sobkulir We have done a number of projects where we essentially do 1b. There is a project-specific mode manager/state manager that components query to see if they can perform various activities. The managers can have arbitrary state machines based on the requirements of the project, so we haven't yet tried to make a standard reusable component that manages state. It is on our roadmap to think about the best way to do that and perhaps incorporate the state checking as an autocoded feature of command handling, but we aren't there yet. 

Typically, you can change state in the following ways (or any combination of these ways):

1) By command. The spacecraft state can be commanded by commands in the state manager. If only these are used, then a reboot can clear the state. The default state can be selected to put the spacecraft in a safe commandable state on boot.

2) By port. The manager can have ports that allow on-board logic to set state. This is a common pattern for fault responses, where on-board detection of a fault condition can put the spacecraft into a "safe" state waiting for ground intervention.

3) By parameter. If the state needs to persist through reboots (intentional or otherwise), a parameter can be set via command and stored via `Svc/PrmDb` for subsequent boot.

The `FpStateMgr` component is part of the picture; it is a way to handle fault and can put the spacecraft into states, but it is not meant as a general state manager.

**Reply by sobkulir** [reply]
This is very helpful @timcanham, thanks! I'm happy to have my line of thought validated :)) Ideally, it would be great to have an open-source example to build upon. But I understand that it's not so simple. There's a mention of HSM framework from JPL in the docs (https://github.com/JPLOpenSource/fprime-scala). Is this the solution you used? It seems to be discontinued so I think the mission used something else. I'll highly appreciate any pointers in this area.
